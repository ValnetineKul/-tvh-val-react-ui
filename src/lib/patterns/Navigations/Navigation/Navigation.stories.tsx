import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import Divider from '../../../components/Divider';
import {
  AngleRight,
  FileCheck,
  Bell,
  Pen,
  FileDownload,
  UserCircle,
  ResearchRequest,
} from '../../../components/Icon/icons/functional';
import Icon from '../../../components/Icon';
import NavigationItem from '../NavigationItem';
import Navigation from './Navigation';
import { createMeta } from '../../../story-utils';

const headerLinks = [
  {
    label: 'Notifications',
    icon: <Icon icon={Bell} />,
    url: '/',
  },
  {
    label: 'My Lists',
    icon: <Icon icon={FileCheck} />,
    url: '/lists',
  },
  {
    label: 'John Doe',
    icon: <Icon icon={UserCircle} />,
    url: '/person',
  },
];

const labelLinks = [
  {
    label: 'Dashboard',
    url: '/',
  },
  {
    label: 'Notifications',
    url: '/notifications',
  },
];

const withStartIconLinks = [
  {
    label: 'My delivery settings with very long menu name which should be truncated',
    icon: <Icon icon={UserCircle} />,
    url: '/contacts',
  },
  {
    label: 'My invoices',
    icon: <Icon icon={Pen} />,
    url: '/invoices',
  },
  {
    label: 'My research results',
    icon: <Icon icon={ResearchRequest} />,
    url: '/results',
  },
];

const withEndIconButtons = [
  {
    label: 'Handling equipment',
    icon: <Icon icon={AngleRight} />,
  },
  {
    label: 'Promotions',
    icon: <Icon icon={AngleRight} />,
  },
];

const withStartAndEndIconButtons = [
  {
    label: 'My payment method',
    startIcon: <Icon icon={UserCircle} />,
    endIcon: <Icon icon={AngleRight} />,
  },
  {
    label: 'My address book',
    startIcon: <Icon icon={FileDownload} />,
    endIcon: <Icon icon={AngleRight} />,
  },
];

const subLabelLinks = [
  {
    label: 'Parts',
    subLabel: 'Extra info',
    url: '/parts',
  },
  {
    label: 'Baskets',
    subLabel: 'All baskets',
    url: '/baskets',
  },
];

export const Primitive = (args: React.ComponentProps<typeof Navigation>) => {
  const [selectedState, setSelectedState] = useState('Notifications');
  const handleClick = (item: string) => {
    selectedState !== item && setSelectedState(item);
  };

  return (
    <Navigation {...args}>
      {headerLinks.map(({ label, icon, url }, key) => {
        return (
          <NavigationItem
            horizontal
            key={key}
            label={label}
            icon={icon}
            href={url}
            selected={selectedState === label}
            onClick={() => handleClick(label)}
          />
        );
      })}
    </Navigation>
  );
};
Primitive.args = {};
Primitive.storyName = 'Horizontal';

export const NavigationVertical = (args: React.ComponentProps<typeof Navigation>) => {
  const [selectedState, setSelectedState] = useState('Notifications');
  const handleClick = (item: string) => {
    selectedState !== item && setSelectedState(item);
  };

  return (
    // Note: -1rem added to delete padding around story
    <div style={{ margin: '-1rem', width: 300 }}>
      <Navigation {...args}>
        {labelLinks.map(({ label, url }, key) => {
          return (
            <NavigationItem
              vertical
              key={key}
              component={Link}
              to={url}
              label={label}
              selected={selectedState === label}
              onClick={() => handleClick(label)}
            />
          );
        })}
        <li role="separator">
          <Divider />
        </li>
        {withStartIconLinks.map(({ label, url, icon }, key) => {
          return (
            <NavigationItem
              vertical
              key={key}
              component={Link}
              to={url}
              startIcon={icon}
              label={label}
              selected={selectedState === label}
              onClick={() => handleClick(label)}
            />
          );
        })}
        <li role="separator">
          <Divider />
        </li>
        {withEndIconButtons.map(({ label, icon }, key) => {
          return (
            <NavigationItem
              vertical
              key={key}
              endIcon={icon}
              label={label}
              selected={selectedState === label}
              onClick={() => handleClick(label)}
            />
          );
        })}
        <li role="separator">
          <Divider />
        </li>
        {subLabelLinks.map(({ label, url, subLabel }, key) => {
          return (
            <NavigationItem
              vertical
              key={key}
              component={Link}
              to={url}
              label={label}
              subLabel={subLabel}
              selected={selectedState === label}
              onClick={() => handleClick(label)}
            />
          );
        })}
        <li role="separator">
          <Divider />
        </li>
        {withStartAndEndIconButtons.map(({ label, startIcon, endIcon }, key) => {
          return (
            <NavigationItem
              vertical
              key={key}
              startIcon={startIcon}
              endIcon={endIcon}
              label={label}
              selected={selectedState === label}
              onClick={() => handleClick(label)}
            />
          );
        })}
      </Navigation>
    </div>
  );
};
NavigationVertical.args = {};
NavigationVertical.storyName = 'Vertical';

export default createMeta({
  component: Navigation,
  title: 'Patterns/Navigations/Navigation',
});
