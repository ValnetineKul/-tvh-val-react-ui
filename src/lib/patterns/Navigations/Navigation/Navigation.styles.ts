import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles({ name: 'Navigation' })((theme) => ({
  root: {},
  list: {
    paddingTop: theme.layoutSpacing['spacing/000'],
    paddingBottom: theme.layoutSpacing['spacing/000'],
  },
}));

export default useStyles;
