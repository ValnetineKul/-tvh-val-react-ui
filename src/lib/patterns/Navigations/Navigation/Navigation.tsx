import type { FC } from 'react';
import React from 'react';

import MenuList from '../../../components/Menus/MenuList';
import useStyles from './Navigation.styles';

export interface NavigationProps {
  className?: string;
}

const Navigation: FC<NavigationProps> = ({ children, className }) => {
  const { classes, cx } = useStyles();

  return (
    <div className={cx(classes.root, className)}>
      <MenuList className={classes.list}>{children}</MenuList>
    </div>
  );
};

export default Navigation;
