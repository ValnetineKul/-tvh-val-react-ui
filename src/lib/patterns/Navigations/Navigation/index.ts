import Navigation from './Navigation';

export type { NavigationProps } from './Navigation';
export default Navigation;
