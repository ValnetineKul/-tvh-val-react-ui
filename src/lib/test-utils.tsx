import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
