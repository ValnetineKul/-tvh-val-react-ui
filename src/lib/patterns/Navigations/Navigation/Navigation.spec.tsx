import React from 'react';
import { render } from '../../../test-utils';
import Navigation from './Navigation';

describe('Navigation', () => {
  it('Should render correctly', () => {
    const { container } = render(<Navigation>Children</Navigation>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
