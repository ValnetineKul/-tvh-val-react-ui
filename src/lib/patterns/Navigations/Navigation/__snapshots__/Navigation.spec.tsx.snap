// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Should render correctly 1`] = `
<div
  class="tss-zxubot-Navigation-root"
>
  <ul
    class="MuiList-root MuiList-padding tss-17qd6i6-Navigation-list mui-h4y409-MuiList-root"
    role="menu"
    tabindex="-1"
  >
    Children
  </ul>
</div>
`;
